src/datumaro/experimental/categories.py

```diff
@@ -131,6 +135,191 @@ def __hash__(self):
+            raise ValueError("Label name cannot be empty and must be a string")
+
+    def __hash__(self):
+        return hash((self.name, self.parent, frozenset(self.label_semantics.items())))
+
+
+@dataclass(frozen=True)
+class LabelGroup:
+    """Represents a group of labels with a specific group type."""
+
+    name: str
+    labels: Tuple[str, ...] = field(default_factory=tuple)
+    group_type: GroupType = GroupType.EXCLUSIVE
+
+    def __post_init__(self):
+        """Validate that name is not empty and labels is a tuple."""
+        if not self.name or not isinstance(self.name, str):
+            raise ValueError("Label group name cannot be empty and must be a string")
+
+
+@dataclass(frozen=True)
+class HierarchicalLabelCategories(Categories):
+    """
+    Represents hierarchical label categories with groups and parent-child relationships.
+    Use this for complex hierarchical classification tasks.
+    """
+
+    items: Tuple[HierarchicalLabelCategory, ...] = field(default_factory=tuple)
+    label_groups: Tuple[LabelGroup, ...] = field(default_factory=tuple)
+    label_semantics: dict = field(default_factory=dict)
+
+    def __post_init__(self):
+        if not isinstance(self.items, tuple):
+            raise TypeError("items must be a tuple of HierarchicalLabelCategory")
+        if not isinstance(self.label_groups, tuple):
+            raise TypeError("label_groups must be a tuple of LabelGroup")
+
+        # Validate no duplicate names
+        seen_names = set()
+        for item in self.items:
+            if item.name in seen_names:
+                raise ValueError(f"Duplicate label name: {item.name}")
+            seen_names.add(item.name)
+
+            # Also check for name label_semantics
+            # TODO: Remove this check after migrating completely to new system
+            if "name" in item.label_semantics:
+                name = item.label_semantics["name"]
+                if name not in seen_names:
+                    seen_names.add(name)
+
+        # Validate that all parents exist
+        for item in self.items:
+            if item.parent and item.parent not in seen_names:
+                raise ValueError(f"Parent '{item.parent}' not found for label '{item.name}'")
+
+        # Validate that all labels in groups exist
+        for group in self.label_groups:
+            for label_name in group.labels:
+                if label_name not in seen_names:
+                    raise ValueError(
+                        f"Label '{label_name}' in group '{group.name}' not found in items"
+                    )
+
+    @property
+    @cache
+    def _index_map(self) -> Dict[str, int]:
+        """Cached mapping from label names to indices."""
+        return {item.name: idx for idx, item in enumerate(self.items)}
+
+    @property
+    @cache
+    def _children_map(self) -> Dict[str, Tuple[str, ...]]:
+        """Cached mapping from parent names to child names."""
+        from collections import defaultdict
```

Comment:
I don't think it will matter if you put this import at the top, I would say this is a bit of an overkill optimization.